src/core/ucc_context.c

```diff
@@ -352,7 +368,71 @@ ucc_status_t ucc_context_destroy(ucc_context_t *context)
         }
         tl_lib->iface->context.destroy(&tl_ctx->super);
     }
+    ucc_progress_queue_finalize(context->pq);
     ucc_free(context->tl_ctx);
     ucc_free(context);
     return UCC_OK;
 }
+
+ucc_status_t ucc_context_progress_register(ucc_context_t *ctx,
+                                           ucc_context_progress_fn_t fn,
+                                           void *progress_arg)
+{
+    int next_pos = ctx->progress_array_size;
+    if (next_pos == ctx->progress_array_max_size) {
+        ctx->progress_array_max_size += 8;
+        ctx->progress_array = ucc_realloc(ctx->progress_array,
```

Comment:
why not to use ucs list instead of array?

Reply:
to have more locality? it is in the most critical path. with the array you fetch 8 pointers (and i don't expect to have more) with 1 cache line. With list you will be going through next into potentially different mem locations.

Reply:
after discussion with Sergey, agreed the benefits are immeasurable. Changed to ucc_list in favor of readability.